File: packages/kotlin/buggie-core/src/main/kotlin/eu/buggie/core/RingBuffer.kt
package eu.buggie.core

/**
 * A fixed-size, thread-safe buffer that keeps the most recent entries.
 *
 * Breadcrumbs and network events are recorded from whatever thread the app happens to
 * be on, and must never grow without bound in a long-running session.
 */
public class RingBuffer<T>(limit: Int) {
    private val limit: Int = maxOf(1, limit)
    private val storage = ArrayDeque<T>()
    private val lock = Any()

    public fun append(element: T) {
        synchronized(lock) {
            storage.addLast(element)

            while (storage.size > limit) storage.removeFirst()
        }
    }

    /** A snapshot. Callers iterate this while other threads are still appending. */
    public val entries: List<T>
        get() = synchronized(lock) { storage.toList() }

    public val count: Int
        get() = synchronized(lock) { storage.size }

    public fun clear() {
        synchronized(lock) { storage.clear() }
    }
}

File: packages/swift/Sources/BuggieCore/RingBuffer.swift
import Foundation

/// A fixed-size, thread-safe buffer that keeps the most recent entries.
///
/// Breadcrumbs and network events are recorded from whatever thread the app happens
/// to be on, and must never grow without bound in a long-running session.
public final class RingBuffer<Element>: @unchecked Sendable {
    private let limit: Int
    private var storage: [Element] = []
    private let lock = NSLock()

    public init(limit: Int) {
        self.limit = max(1, limit)
        storage.reserveCapacity(self.limit)
    }

    public func append(_ element: Element) {
        lock.lock()
        defer { lock.unlock() }

        storage.append(element)

        if storage.count > limit {
            storage.removeFirst(storage.count - limit)
        }
    }

    public var entries: [Element] {
        lock.lock()
        defer { lock.unlock() }

        return storage
    }

    public var count: Int {
        lock.lock()
        defer { lock.unlock() }

        return storage.count
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }

        storage.removeAll(keepingCapacity: true)
    }
}
